### PR TITLE
Rewrite introduction

### PR DESCRIPTION
**Pull request by adamgundry**
Based on comments from others at Well-Typed, I've rewritten the introduction in the `Optics` main module. In particular it now actually explains what optics are. Further feedback welcome.

### CODE REVIEW

File: optics/src/Optics.hs [hunk 1]
```diff
@@ -224,52 +199,356 @@ import Data.Either.Optics                    as P
...
+--
+-- == Advantages of the opaque design
+--
+-- Since the interface is deliberately chosen rather than to some extent
+-- determined by the implementation, we are free to choose a more restricted
+-- interface where doing so leads to conceptual simplicity.  For example, in
+-- @lens@, the 'view' function can be used with a 'Fold' provided the result
+-- type has a 'Monoid' instance, and the multiple targets of the 'Fold' will be
+-- combined monoidally.  This behaviour can be confusing, so in @optics@ a
+-- 'Getter' cannot be silently used as a 'Fold', and we prefer to have 'view'
```
Comment (arybczak) [inline optics/src/Optics.hs]: I think you meant the other way around, i.e. "a Fold cannot be silently used as a Getter".

**Comment by arybczak** [comment]
Thanks :+1: Apart from my single comment it LGTM.

File: optics/src/Optics.hs (hunk 1 above)
Comment (adamgundry) [reply optics/src/Optics.hs]: Well spotted! Fixed.

**Comment by adamgundry** [comment]
I've added lots of other minor doc improvements to this PR, mostly to minimise merge conflicts.

**Comment by arybczak** [comment]
Shall we merge then?

**Pull request merged**